app/src/main/java/com/zeapo/pwdstore/pwgen/PRNGFixes.kt

```diff
@@ -0,0 +1,326 @@
+package com.zeapo.pwdstore.pwgen
+
+/*
+ * This software is provided 'as-is', without any express or implied
+ * warranty.  In no event will Google be held liable for any damages
+ * arising from the use of this software.
+ *
+ * Permission is granted to anyone to use this software for any purpose,
+ * including commercial applications, and to alter it and redistribute it
+ * freely, as long as the origin is not misrepresented.
+ */
+
+import android.os.Build
+import android.os.Process
+import android.util.Log
+
+import java.io.ByteArrayOutputStream
+import java.io.DataInputStream
+import java.io.DataOutputStream
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.io.IOException
+import java.io.OutputStream
+import java.io.UnsupportedEncodingException
+import java.security.NoSuchAlgorithmException
+import java.security.Provider
+import java.security.SecureRandom
+import java.security.SecureRandomSpi
+import java.security.Security
+
+/**
+ * Fixes for the output of the default PRNG having low entropy.
+ *
+ * The fixes need to be applied via [.apply] before any use of Java
+ * Cryptography Architecture primitives. A good place to invoke them is in the
+ * application's `onCreate`.
+ */
+object PRNGFixes {
+
+    private const val VERSION_CODE_JELLY_BEAN = 16
+    private const val VERSION_CODE_JELLY_BEAN_MR2 = 18
+    private val BUILD_FINGERPRINT_AND_DEVICE_SERIAL = buildFingerprintAndDeviceSerial
+
+    /**
+     * Gets the hardware serial number of this device.
+     *
+     * @return serial number or `null` if not available.
+     */
+    private// We're using the Reflection API because Build.SERIAL is only available
```

Comment:
The project has a minimum of api level 16. Wouldn't that simplify this?

Reply:
It will, but `Build.SERIAL` is deprecated and `Build.getSerial()` which replaces it requires `android.permission.READ_PHONE_STATE` so I'm not really sure what path to take here.

Reply:
Would not make sens to request `READ_PHONE_STATE` indeed

Reply:
also from the doc, the value of this field is not `UNKOWN`:
> For apps targeting SDK higher than Build.VERSION_CODES.O_MR1 this field is set to UNKNOWN.

Reply:
I guess so. From my readings the issue was present only on API 16, for cryptographic reasons it is important. We're using it only for generating random passwords (not random enough on API 16).
We currently have only tens (less then 20 AFAIK) users on the Play Store running on 4.1-4.2. 

Removing this part would remove a bunch of code, less maintenance :)

Reply:
Awesome. Whenever this gets merged I'll bump minSdk to something reasonable like 19 or 21 and do a full clean-up.

Reply:
Let me known when you think that this PR is done (with the previous code removed). I'll merge it :)

Reply:
It's done from my end.

Reply:
Thanks. Can you resolve the conflict below please?
I'll merge it afterwards.